nx/lib/nx.ex

```diff
@@ -6142,15 +6142,75 @@ defmodule Nx do
         1
       >
 
+  Although `NaN` by definition isn't equal to itself, this implementation
+  considers all `NaN`s equal to each other and only to each other:
+
+      iex> Nx.all_close(Nx.tensor(:nan), Nx.tensor(:nan))
+      #Nx.Tensor<
+        u8
+        1
+      >
+
+      iex> Nx.all_close(Nx.tensor(:nan), Nx.tensor(0))
+      #Nx.Tensor<
+        u8
+        0
+      >
+
+  Infinities behave as expected, being "close" to themselves but not
+  to other numbers:
+
+      iex> Nx.all_close(Nx.tensor(:infinity), Nx.tensor(:infinity))
+      #Nx.Tensor<
+        u8
+        1
+      >
+
+      iex> Nx.all_close(Nx.tensor(:infinity), Nx.tensor(:neg_infinity))
+      #Nx.Tensor<
+        u8
+        0
+      >
+
+      iex> Nx.all_close(Nx.tensor(1.0e30), Nx.tensor(:infinity))
+      #Nx.Tensor<
+        u8
+        0
+      >
   """
   @doc type: :aggregation
   def all_close(a, b, opts \\ []) do
     opts = keyword!(opts, rtol: 1.0e-5, atol: 1.0e-8)
     rtol = opts[:rtol]
     atol = opts[:atol]
 
-    # TODO: deal with non_finite entries by adding is_infinity and is_nan
-    all(less_equal(Nx.abs(subtract(a, b)), add(atol, multiply(rtol, Nx.abs(b)))))
+    a = to_tensor(a)
+    b = to_tensor(b)
+
+    nan_a = is_nan(a)
+    nan_b = is_nan(b)
+    nan_selector = logical_or(nan_a, nan_b)
+    nan_entries = logical_and(nan_a, nan_b)
+
+    inf_a = is_infinity(a)
+    inf_b = is_infinity(b)
+    inf_selector = logical_or(inf_a, inf_b)
```

Comment:
I don't think we need to select infinity, because it will still compare property. I think we want something like this:

    mask_a = is_nan(a)
    mask_b = is_nan(b)
    mask_a == mask_b and select(mask_a, 0, a) == select(mask_b, 0, b)

Reply:
This and `:nan - :nan === :nan` were the reasons for adding is_nan and is_infinity actually